### PR TITLE
Fix predicate return

### LINKED ISSUE #3086: Predicate function retrun value is not compatiable with old plugins
There are other plugins dynamically loaded to volcano. eg. huawei's acend gpu, https://gitee.com/ascend/ascend-for-volcano

**What happened**:

https://github.com/volcano-sh/volcano/blob/b95d0f15e0ff2af9a7c9ae9a037a4c4c9e7300aa/pkg/scheduler/api/types.go#L170C6-L170C17
```
// PredicateFn is the func declaration used to predicate node for task.
type PredicateFn func(*TaskInfo, *NodeInfo) ([]*Status, error)
```

npu plugin
https://gitee.com/ascend/ascend-for-volcano/blame/master/huawei_npu.go
```
	// if npu no meet the task require,the task will failed.so need to intercept in advance
	ssn.AddPredicateFn(tp.Name(), func(taskInfo *api.TaskInfo, nodeInfo *api.NodeInfo) error {
		return sHandler.NodePredicate(taskInfo, nodeInfo, ssn.Configurations)
	})
```

**What you expected to happen**:

Should be compatiable with other extender plugins

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

Introduced by refact PR [#2916](https://github.com/volcano-sh/volcano/pull/2916)

**Environment**:
- Volcano Version: master
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by lowang-bh**
fixes #3086

Part-2 of [filter out those nodes which are UnschedulableAndUnresolvable when preempting #3432](https://github.com/volcano-sh/volcano/pull/3432)
Please merge 3432 first.

<details>

<summary> pod messages examples </summary>

image
</details>

### CODE REVIEW

File: example/extender/extender.go [hunk 1]
```diff
@@ -72,10 +72,7 @@ func predicate(w http.ResponseWriter, r *http.Request) {
 
 	resp := &extender.PredicateResponse{}
 	if req.Task.BestEffort && len(req.Node.Tasks) > 10 {
-		sts := api.Status{}
-		sts.Code = api.Unschedulable
-		sts.Reason = "Too many tasks on the node"
-		resp.Status = append(resp.Status, &sts)
+		resp.ErrorMessage = "Too many tasks on the node"
```
Comment (Monokaix) [inline example/extender/extender.go:75]: Do we also need return statuscode？

File: pkg/scheduler/framework/session_plugins.go [hunk 2]
```diff
@@ -638,14 +637,59 @@ func (ssn *Session) PredicateFn(task *api.TaskInfo, node *api.NodeInfo) ([]*api.
...
 	}
-	return predicateStatus, nil
+	return nil
+}
+
+// UnscheduleAndUnresolvedError checks if the predicate error contains
+// - Unschedulable
+// - UnschedulableAndUnresolvable
+// - ErrorSkipOrWait
+func (ssn *Session) UnscheduleAndUnresolvedError(task *api.TaskInfo, node *api.NodeInfo) error {
```
Comment (Monokaix) [inline pkg/scheduler/framework/session_plugins.go:653]: Put these to file `session.go` is better.

File: pkg/scheduler/actions/allocate/allocate.go [hunk 3]
```diff
@@ -303,23 +303,14 @@ func (alloc *Action) allocateResourcesForTasks(tasks *util.PriorityQueue, job *a
...
-	if err != nil {
-		return nil, api.NewFitError(task, node, err.Error())
-	}
-
-	if statusSets.ContainsUnschedulable() || statusSets.ContainsUnschedulableAndUnresolvable() ||
-		statusSets.ContainsErrorSkipOrWait() {
-		return nil, api.NewFitErrWithStatus(task, node, statusSets...)
-	}
-	return nil, nil
+	return alloc.session.UnscheduleAndUnresolvedError(task, node)
```
Comment (Monokaix) [inline pkg/scheduler/actions/allocate/allocate.go:313]: I think it's better to separate predictions according to actions, so `UnscheduleAndUnresolvedError` can change to something like `predictWhenAllocate`, and so dose `UnresolvedError`.

Comment (Monokaix) [reply pkg/scheduler/actions/allocate/allocate.go:313]: current func name is not readable.

Comment (lowang-bh) [reply pkg/scheduler/actions/allocate/allocate.go]: Done.

File: pkg/scheduler/framework/session_plugins.go (hunk 2 above)
Comment (lowang-bh) [reply pkg/scheduler/framework/session_plugins.go]: Done.

File: example/extender/extender.go (hunk 1 above)
Comment (lowang-bh) [reply example/extender/extender.go:75]: The default code is `error`.

**Comment by Monokaix** [comment]
When upgrade from 1.8/1.9 to 1.10 seems compatibility issue wil also occur: )

**Comment by lowang-bh** [comment]
> When upgrade from 1.8/1.9 to 1.10 seems compatibility issue wil also occur: )

Currently, no body has reported the issue after release 1.8/1.9. So maybe they did't met this issue with self-developed plugins.

File: pkg/scheduler/api/types.go [hunk 4]
```diff
@@ -227,6 +228,30 @@ func (s StatusSets) Reasons() []string {
 	return all
 }
 
+// ConvertPredicateStatus return predicate status from k8sframework status
+func ConvertPredicateStatus(status *k8sframework.Status) *Status {
+	internalStatus := &Status{}
+	if status != nil {
+		internalStatus.Plugin = status.Plugin() // function didn't check whether Status is nil
+	}
+	if status.Code() == k8sframework.Success {
```
Comment (googs1025) [inline pkg/scheduler/api/types.go:237]: Will it panic if `status == nil`?

File: pkg/scheduler/api/types.go [hunk 5]
```diff
@@ -227,6 +228,30 @@ func (s StatusSets) Reasons() []string {
...
+		return internalStatus
+	} else if status.Code() == k8sframework.UnschedulableAndUnresolvable {
+		internalStatus.Code = UnschedulableAndUnresolvable
+		internalStatus.Reason = status.Message()
+		return internalStatus
+	} else {
+		internalStatus.Code = Error
+		internalStatus.Reason = status.Message()
+		return internalStatus
+	}
```
Comment (googs1025) [inline pkg/scheduler/api/types.go:252]: use switch is better? 
```go
          switch status.Code() {
	case k8sframework.Success:
		internalStatus.Code = Success
	case k8sframework.Unschedulable:
		internalStatus.Code = Unschedulable
	case k8sframework.UnschedulableAndUnresolvable:
		internalStatus.Code = UnschedulableAndUnresolvable
	default:
		internalStatus.Code = Error
	}

	internalStatus.Reason = status.Message()
	return internalStatus


```

File: pkg/scheduler/api/types.go (hunk 4 above)
Comment (lowang-bh) [reply pkg/scheduler/api/types.go]: No

File: example/extender/extender.go (hunk 1 above)
Comment (lowang-bh) [reply example/extender/extender.go:75]: Have added a code field in respond.

File: pkg/scheduler/plugins/predicates/predicates.go [hunk 6]
```diff
@@ -432,31 +432,34 @@ func (pp *predicatesPlugin) OnSessionOpen(ssn *framework.Session) {
...
 		if predicate.nodePortEnable {
 			isSkipNodePorts := handleSkipPredicatePlugin(task, skipPlugins, nodePortFilter.Name(), node)
 			if !isSkipNodePorts {
 				status := nodePortFilter.Filter(context.TODO(), state, nil, nodeInfo)
-				nodePortStatus := framework.ConvertPredicateStatus(status)
+				nodePortStatus := api.ConvertPredicateStatus(status)
 				if nodePortStatus.Code != api.Success {
 					predicateStatus = append(predicateStatus, nodePortStatus)
-					return predicateStatus, fmt.Errorf("plugin %s predicates failed %s", nodePortFilter.Name(), status.Message())
+					return api.NewFitErrWithStatus(task, node, predicateStatus...)
```
Comment (wangyang0616) [inline pkg/scheduler/plugins/predicates/predicates.go:462]: When multiple filtering algorithms are executed in the predicate plug-in, for example, nodePortFilter, podAffinityFilter, volumeZoneFilter, and podTopologySpreadFilter,

If the result of nodePortFilter is Unschedulable, the result of podAffinityFilter is UnschedulableAndUnresolvable. The predicate plug-in is returned after nodePortFilter is executed. Other algorithm plug-ins are not executed.

When the preempt action is used to determine whether the current node meets the conditions, the corresponding status code contains only the nodePortFilter information and is Unschedulable, but does not contain the UnschedulableAndUnresolvable information of podAffinityFilter. In this case, the system considers that the node meets the preemption conditions, resulting in misjudgment.

Of course, this is an existing problem, not introduced by this pr, but we can discuss the idea together.

File: pkg/scheduler/plugins/usage/usage.go [hunk 7]
```diff
@@ -124,18 +123,17 @@ func (up *usagePlugin) OnSessionOpen(ssn *framework.Session) {
...
 		predicateStatus := make([]*api.Status, 0)
-		usageStatus := &api.Status{}
+		usageStatus := &api.Status{Plugin: PluginName}
 
 		now := time.Now()
 		if up.period == "" || now.Sub(node.ResourceUsage.MetricsTime) > MetricsActiveTime {
 			klog.V(4).Infof("The period(%s) is empty or the usage metrics data is not updated for more than %v minutes, "+
 				"Usage plugin filter for task %s/%s on node %s pass, metrics time is %v. ", up.period, MetricsActiveTime, task.Namespace, task.Name, node.Name, node.ResourceUsage.MetricsTime)
 
 			usageStatus.Code = api.Success
```
Comment (wangyang0616) [inline pkg/scheduler/plugins/usage/usage.go:135]: This line can be deleted.

File: pkg/scheduler/plugins/predicates/predicates.go (hunk 6 above)
Comment (lowang-bh) [reply pkg/scheduler/plugins/predicates/predicates.go:462]: That is a problem in current architecture.  We have discussed this problem with @Monokaix and we can following the kubernetes process. And the preemtion/reclaim action can also use the predicating result cache.

I think we can record a issue under https://github.com/volcano-sh/volcano/issues/3502.

File: pkg/scheduler/plugins/usage/usage.go (hunk 7 above)
Comment (lowang-bh) [reply pkg/scheduler/plugins/usage/usage.go]: Done.

File: pkg/scheduler/framework/session.go [hunk 8]
```diff
@@ -313,6 +313,52 @@ func (ssn *Session) GetUnschedulableAndUnresolvableNodesForTask(task *api.TaskIn
 	return ret
 }
 
+// PredicateWhenAllocate checks if the predicate error contains
+// - Unschedulable
+// - UnschedulableAndUnresolvable
+// - ErrorSkipOrWait
+func (ssn *Session) PredicateWhenAllocate(task *api.TaskInfo, node *api.NodeInfo) error {
```
Comment (william-wang) [inline pkg/scheduler/framework/session.go:320]: Overall looks good to me. Suggest to chang name to `PredicateForAllocateAction`

File: pkg/scheduler/framework/session.go [hunk 9]
```diff
@@ -313,6 +313,52 @@ func (ssn *Session) GetUnschedulableAndUnresolvableNodesForTask(task *api.TaskIn
...
+		statusSets.ContainsErrorSkipOrWait() {
+		return fitError
+	}
+	return nil
+}
+
+// PredicateWhenPreempt checks if the predicate error contains:
+// - UnschedulableAndUnresolvable
+// - ErrorSkipOrWait
+func (ssn *Session) PredicateWhenPreempt(task *api.TaskInfo, node *api.NodeInfo) error {
```
Comment (william-wang) [inline pkg/scheduler/framework/session.go:342]: Overall looks good to me. Suggest to chang name to PredicateForPreemptAction

**Review by william-wang — APPROVED** [review]
/approve

**Pull request merged**